src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -385,10 +383,299 @@ public static void EnsureSocketsAreInitialized()
+                if (!SetCallbackStartedOrCanceled())
+                    return;
+
+                ReleaseResources(cancelQuery: true);
+            }
+
+            [ReliabilityContract(Consistency.WillNotCorruptState, Cer.Success)]
+            public void Dispose()
+            {
+                GC.SuppressFinalize(this);
+                ReleaseResources(cancelQuery: false);
+            }
+
+            [ReliabilityContract(Consistency.WillNotCorruptState, Cer.Success)]
+            private void ReleaseResources(bool cancelQuery)
+            {
+                var ptr = Interlocked.Exchange(ref m_context, IntPtr.Zero);
+
+                if (ptr == IntPtr.Zero)
+                    return;
+
+                var context = (GetAddrInfoExContext*)ptr;
+
+                if (cancelQuery)
+                    context->CancelQuery();
+
+                GetAddrInfoExContext.FreeContext(context);
+            }
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        private unsafe struct GetAddrInfoExContext
+        {
+            private static readonly int Size = Marshal.SizeOf<GetAddrInfoExContext>();
+
+            public NativeOverlapped Overlapped;
+            public AddressInfoEx* Result;
+            public IntPtr QueryStateHandle;
+            public IntPtr CancelHandle;
+
+            public GetAddrInfoExContext(GetAddrInfoExState state)
+            {
+                Overlapped = new NativeOverlapped();
+                Result = null;
+
+                var handle = GCHandle.Alloc(state, GCHandleType.Normal);
+                QueryStateHandle = GCHandle.ToIntPtr(handle);
+
+                CancelHandle = IntPtr.Zero;
+            }
+
+            public GetAddrInfoExState GetQueryState()
+            {
+                var stateHandle = Interlocked.Exchange(ref QueryStateHandle, IntPtr.Zero);
+
+                if (stateHandle == IntPtr.Zero)
+                    return null;
+
+                var handle = GCHandle.FromIntPtr(stateHandle);
+
+                if (!handle.IsAllocated)
+                    return null;
+
+                var state = (GetAddrInfoExState)handle.Target;
+                handle.Free();
+
+                return state;
+            }
+
+            [ReliabilityContract(Consistency.WillNotCorruptState, Cer.Success)]
+            public void CancelQuery()
+            {
+                if (CancelHandle != IntPtr.Zero)
+                {
+                    Interop.Winsock.GetAddrInfoExCancel(ref CancelHandle);
+                    CancelHandle = IntPtr.Zero;
+                }
+            }
+
+            public static GetAddrInfoExContext* AllocateContext(string hostAddress, DnsResolveAsyncResult asyncResult)
+            {
+                GetAddrInfoExContext* context = (GetAddrInfoExContext*)Marshal.AllocHGlobal(Size);
+                GetAddrInfoExState state = new GetAddrInfoExState(hostAddress, asyncResult);
+
+                *context = new GetAddrInfoExContext(state);
```

Comment:
This seems a little dangerous.  We're computing how much space to allocate based on `Marshal.SizeOf<GetAddrInfoExContext>` and then doing a native allocation of that size, but then we're storing a managed GetAddrInfoExContext into that space.  This assumes that the managed and native sizes are the same.  Presumably they are, or else other stuff would break, but it'd be good to validate this.  Can you add an assert that `Unsafe.SizeOf<GetAddrInfoExContext>() == Size`?